contributors/design-proposals/node/numa-manager.md

```diff
@@ -0,0 +1,281 @@
+
+Note that all of these concerns pertain only to multi-socket systems.
+
+## Goals
+
+- Allow CPU manager and Device Manager to agree on preferred
+  NUMA node affinity for containers.
+- Provide an internal interface and pattern to integrate additional
+  topology-aware Kubelet components.
+
+## Non-Goals
+
+- _Inter-device connectivity:_ Decide device assignments based on direct
+  device interconnects. This issue can be separated from NUMA node
+  locality. Inter-device topology can be considered entirely within the
+  scope of the Device Manager, after which it can emit possible
+  NUMA affinities. The policy to reach that decision can start simple
+  and iterate to include support for arbitrary inter-device graphs.
+- _HugePages:_ This proposal assumes that pre-allocated HugePages are
+  spread among the available NUMA nodes in the system. We further assume
+  the operating system provides best-effort local page allocation for
+  containers (as long as sufficient HugePages are free on the local NUMA
+  node.
+- _CNI:_ Changing the Container Networking Interface is out of scope for
+  this proposal. However, this design should be extensible enough to
+  accommodate network interface locality if the CNI adds support in the
+  future. This limitation is potentially mitigated by the possiblity to
+  use the device plugin API as a stopgap solution for specialized
+  networking requirements.
+
+## User Stories
+
+*Story 1: Fast virtualized network functions*
+
+A user asks for a "fast network" and automatically gets all the various
+pieces coordinated (hugepages, cpusets, network device) co-located on a
+NUMA node.
+
+*Story 2: Accelerated neural network training*
+
+A user asks for an accelerator device and some number of exclusive CPUs
+in order to get the best training performance, due to NUMA-alignment of
+the assigned CPUs and devices.
+
+# Proposal
+
+*Main idea: Two Phase NUMA coherence protocol*
+
+NUMA affinity is tracked at the container level, similar to devices and
+CPU affinity. At pod admission time, a new component called the NUMA Manager
+collects possible NUMA configurations from the Device Manager and the
+CPU Manager. The NUMA manager acts as an oracle for NUMA node affinity by
+those same components when they make concrete resource allocations. We
+expect the consulted components to use the inferred QoS class of each
+pod in order to prioritize the importance of fulfilling optimal NUMA
+affinity.
+
+## Proposed Changes
+
+### New Component: NUMA Manager
+
+This proposal is focused on a new component in the Kubelet called the
+NUMA Manager. The NUMA Manager implements the pod admit handler
+interface and participates in Kubelet pod admission. When the `Admit()`
+function is called, the NUMA manager collects NUMA hints from from other
+Kubelet components.
+
+If the NUMA hints are not compatible, the NUMA manager could choose to
+reject the pod. The details of what to do in this situation needs more
```

Comment:
According to [the docs](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1117) if scheduler is unaware of the rejection rule then `softAdmitHandler` should be used. This is the case described here so any pod rejected at this phase will stay in `Pending` state indefinitely. 
Suggesting user to utilize a single-pod ReplicaSet (or any other controller capable of deleting and re-creating a pod) might be a good idea. Otherwise user might not be aware of failed scheduling.

Reply:
Great points, will add.